Bluetooth: GATT: Fix usage of bt_gatt_foreach_attr_type

Vudentz opened the pull request:
bt_gatt_foreach_attr_type make use of stack variable to store static
service attributes so the handle could be resolved (as static services
are read-only), but code such as bt_gatt_indicate and bt_gatt_notify
use it out of scope when an UUID is provided.

joerchan commented:
I don't see it how the invalid read happened, at the moment the callback is called the stack variable should be valid.
Is the issue that a reference is kept to the the attribute, in that case this wouldn't solve it, because while the memory location might be valid, it might have been copied over from another iteration.

Thalley commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -1474,7 +1474,7 @@ void bt_gatt_foreach_attr_type(uint16_t start_handle, uint16_t end_handle,
 
 			for (i = 0; i < static_svc->attr_count; i++, handle++) {
-				struct bt_gatt_attr attr;
+				static struct bt_gatt_attr attr;
```
I don't understand why this fixes the issue (I'm assuming that you've tested with valgrind to validate this fix). Can you elaborate? 

I also wonder that if this fixes the issue here, do we need a similar addition of `static` to `foreach_attr_type_dyndb`?

Vudentz commented:
> I don't see it how the invalid read happened, at the moment the callback is called the stack variable should be valid.
> Is the issue that a reference is kept to the the attribute, in that case this wouldn't solve it, because while the memory location might be valid, it might have been copied over from another iteration.

Maybe my explanation is not proper, there are actually code that would attempt to access the pointer outside the scope:

https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/bluetooth/host/gatt.c#L2065
https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/bluetooth/host/gatt.c#L2152

So both notify and indicate do actually endup using the attribute past its scope, we could however make yet another variable to store the contents but in case the attribute happens to be in the dynamic area then this would just cause extra copies, another option would be to have the handle be resolved for static services but return the original attribute pointer and in addtion the handle, though would break the current API so I just made it static which which perhaps is a better fix after all.

joerchan commented:
@Vudentz I don't see the use of static as a valid solution. You linked to places in the API which would take an attr pointer, but this is given by the application. Which means that we don't have control over the time between when the foreach function was called, and when this was given as an input to the API, there is no a race-condition on this static memory location, it's contents can be overwritten at any time.

The API's you linked to would handle the case where the pointer was a static one i.e the handle is set to zero.
I think the best would be to give the original pointer, and provide access function for getting the handle, so that the application can get the static handle as well.
The other option I see is to actually make it so that the static attributes have valid handles, although that might be hard to do and could require involving the build system.

Vudentz commented:
> @Vudentz I don't see the use of static as a valid solution. You linked to places in the API which would take an attr pointer, but this is given by the application. Which means that we don't have control over the time between when the foreach function was called, and when this was given as an input to the API, there is no a race-condition on this static memory location, it's contents can be overwritten at any time.

Yep, the use of static variable could be racy indeed, so Ive done the fix using a temporary variable that just stores the attribute at the correct scope.

> The API's you linked to would handle the case where the pointer was a static one i.e the handle is set to zero.
> I think the best would be to give the original pointer, and provide access function for getting the handle, so that the application can get the static handle as well.
> The other option I see is to actually make it so that the static attributes have valid handles, although that might be hard to do and could require involving the build system.

Well if we provide the original pointer it would break the API anyway since it will no longer have the handle resolved and we would depend on yet another API call to resolve it which would likely break application that did use these APIs in the past, so like I said if we could break the API IVe just add another argument to the callback passing the resolved handle along with the original attribute pointer.

Attempt to fill up the handles at build-time maybe possible but the problem is that the sections are created at linking time so we would need to loop thru every attribute on each service filling its handles.

Thalley commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -2054,18 +2068,12 @@ int bt_gatt_notify_cb(struct bt_conn *conn,
-
-		bt_gatt_foreach_attr_type(handle, 0xffff, params->uuid,
-					  NULL, 1, match_uuid, &attr);
+		attr = gatt_find_by_uuid(&tmp, handle, params->uuid);
```
It's a static function, so perhaps it's not so bad, but having an API where you provide a pointer to temporary value to a function, without using the result of said pointer, seems somewhat odd. 

But I guess that in this case, `attr == &tmp`, right? So I guess the `attr` parameter to `gatt_find_by_uuid` is basically an `[out]` parameter, and the return value is basically just a nice-to-have thing.

Thalley reviewed (dismissed):
One comment (doesn't need any changes, but perhaps a confirmation/reply?). Otherwise LGTM assuming that the invalid read is properly dealt with.

joerchan commented:
> Well if we provide the original pointer it would break the API anyway since it will no longer have the handle resolved and we would depend on yet another API call to resolve it which would likely break application that did use these APIs in the past, so like I said if we could break the API IVe just add another argument to the callback passing the resolved handle along with the original attribute pointer.

The application have access to the pointer directly, so it might be exposed to the problem of the handle being zero from that,
we might need the handle resolving function for that.
Also the attribute is not temporary only in the foreach callback, but also the read and write callbacks of the attribute, so it would have to be documented for this one as well.

Also `bt_gatt_attr_next` returns an invalid pointer.

The application is exposed to both the temporary pointer, and the zero handle, and as far as I can see no good way to tell if the pointer is temporary or not.

I think we should also create a task to get the handles assigned for the static attributes, so that we can remove the use of temporary pointer.

joerchan commented:
Alternative solution for no build handle resolution could be to make static defines a Kconfig. And put them back into RAM, by default.
Then those that enable the static attributes will have to use the handle resolution function. That would not be an API break in that case.

asbjornsabo commented on include/bluetooth/gatt.h:
```diff
@@ -338,6 +338,10 @@ enum {
 /** @typedef bt_gatt_attr_func_t
  *  @brief Attribute iterator callback.
  *
+ *  Note: Attributes passed to this callback are temporary when it happens to
```
This may be due to a lack of knowledge of gatt.c, but I do not fully understand this comment.
I think "Attributes passed to this callback are temporary if the attribute happens to be in the static portion of the database, so the attribute can not be used as the original location" is clearer.  (Gets rid of the "it" to refer to the "attributes" in plural.)
But "can not be used the original location" is still not clear to me.  (I understand that it is temporary, and therefore should not be used.  but as the "original location"?)

Vudentz commented:
> > Well if we provide the original pointer it would break the API anyway since it will no longer have the handle resolved and we would depend on yet another API call to resolve it which would likely break application that did use these APIs in the past, so like I said if we could break the API IVe just add another argument to the callback passing the resolved handle along with the original attribute pointer.
> 
> The application have access to the pointer directly, so it might be exposed to the problem of the handle being zero from that,
> we might need the handle resolving function for that.

And then introduce the anti pattern of double lookup (since handle resolutation would also lookup), which is why I said that if we do want to pass the original pointer back then we should just resolve in place and give the handle back to the callback.

> Also the attribute is not temporary only in the foreach callback, but also the read and write callbacks of the attribute, so it would have to be documented for this one as well.
> 
> Also `bt_gatt_attr_next` returns an invalid pointer.
> 
> The application is exposed to both the temporary pointer, and the zero handle, and as far as I can see no good way to tell if the pointer is temporary or not.
> 
> I think we should also create a task to get the handles assigned for the static attributes, so that we can remove the use of temporary pointer.

Sure, it just doesn't seem there is an easy way to do it.

joerchan commented:
> Sure, it just doesn't seem there is an easy way to do it.

True, but seems like the best solution. Something like how offsets.c is generated or SW ISR table, might not end up being pretty, in the build step. But nice from the API perspective.

> Also bt_gatt_attr_next returns an invalid pointer.

This one should be fixed though.

Vudentz committed:
Bluetooth: GATT: Optimize static service handle resolution

This optimizes the lookups to skip ahead when attribute pointer is not
located within service array.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: GATT: Rework bt_gatt_attr_func_t

Make it actually give the original pointer to the attribute and its
resolved handle so static attributes don't need an extra lookup.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: GATT: Introduce bt_gatt_attr_get_handle

This introduces bt_gatt_attr_get_handle which can be used to resolve
handles of static attributes.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz commented:
> > Sure, it just doesn't seem there is an easy way to do it.
> 
> True, but seems like the best solution. Something like how offsets.c is generated or SW ISR table, might not end up being pretty, in the build step. But nice from the API perspective.

Ive looking into it as well, but it may take some time to make something clean enough so we can store the resolve values in rom.

> > Also bt_gatt_attr_next returns an invalid pointer.
> 
> This one should be fixed though.

Sure, Ive actually have it fixed now but it breaks the API to return the original pointer + handle so at least we can use this until we figure out how to fix the values on rom.

carlescufi commented on include/bluetooth/gatt.h:
```diff
@@ -339,12 +339,14 @@ enum {
  *  @return BT_GATT_ITER_STOP to stop.
  */
 typedef uint8_t (*bt_gatt_attr_func_t)(const struct bt_gatt_attr *attr,
+				       uint16_t handle,
```
Please add a note in the release notes, API changes section.

Vudentz replied:
Sure, where is it located on tree?

Vudentz replied:
Pushed an update including some text in API changes for bt_gatt_attr_func_t.

Thalley commented on include/bluetooth/gatt.h:
```diff
@@ -392,6 +394,15 @@ static inline void bt_gatt_foreach_attr(uint16_t start_handle, uint16_t end_hand
 
+/** @brief Get Attribute handle.
+ *
+ * @param attr Attribute object.
+ *
+ * @return Haandle of the corresponding attribute or zero if the attribute
+ * could not be found.
```
Missing a space to align with the @brief

Vudentz replied:
Will fix it.

Vudentz commented on include/bluetooth/gatt.h:
```diff
@@ -392,6 +394,15 @@ static inline void bt_gatt_foreach_attr(uint16_t start_handle, uint16_t end_hand
+ *
+ * @param attr Attribute object.
+ *
+ * @return Haandle of the corresponding attribute or zero if the attribute
```
Will fix it.

Vudentz commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -1268,16 +1270,31 @@ static uint8_t get_service_handles(const struct bt_gatt_attr *attr,
+	}
+
 	Z_STRUCT_SECTION_FOREACH(bt_gatt_service_static, static_svc) {
+		/* Skip ahead if start is not within service attributes array */
```
Afaik the sections are sorted alphabetically, so the services pointers may not be in ascending order thus why Im checking the full range, that said it should be indeed static_svc->attr_count - 1 for the end range.

joerchan replied:
I can see that, looks like for native_posix they are in descending order, and that thing is what is failing the bt_bsim test.
This check does not seem to work in that case.

Commenting out the skip-ahead part then this works as expected.

Vudentz commented:
Took me a lot more time to check what is going on with test then the actual problem.

Summary:

Status   Count
==============
PASS        69
==============
Total       69

Vudentz commented:
@carlescufi Looks like it is finally passing all tests.
